A22-gssapi-authentication.md

```diff
@@ -0,0 +1,83 @@
+Add Optional GSSAPI call credentials implementation
+---------------------------------------------------
+
+* Author: David Li (@lihalite)
+* Approver: a11r
+* Status: Draft
+* Implemented in: C++
+* Last updated: 2018/09/19
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/8ox3Nu-4b4o
+
+## Abstract
+
+gRPC client libraries will include an optional call credentials that
+uses GSSAPI to acquire a token, added to the metadata on each call.
+
+## Background
+
+gRPC client libraries provide an authentication API based on a
+Credential interface, as well as a set of Credential implementations
+for common authentication methods including TLS and OAuth2. Another
+authentication method, currently unimplemented, is the Kerberos
+protocol, generally used through the GSSAPI standard, and often found
+in corporate environments.
+
+### Related Proposals:
+
+N/A
+
+## Proposal
+
+gRPC will optionally support GSSAPI authentication via a provided
+Credential implementation. To avoid introducing a new hard dependency
+for everyone, GSSAPI support will only be compiled/provided when
+configured at build time. Since multiple GSSAPI implementations exist,
+the user can specify the implementation at build time, by providing
+the library to link to and any necessary include paths (for C++).
+
+If not enabled at build time, then the implementation will be a stub
+that simply logs a message at the INFO warning level.
+
+In grpc_cli, if enabled at build time, we will create a composite
+channel credentials that includes the GSSAPI call credentials, so that
+any requests made are authenticated.
+
+## Rationale
+
+By making the dependency on GSSAPI optional, we avoid introducing new
+hard dependencies for downstream users, but we also make the build
+process more complex. However, by placing it in the C++ core, we can
+integrate it with grpc_cli, allowing it to be used to debug and test
+gRPC servers in environments requiring GSSAPI authentication.
+
+Alternate approaches:
+
+The Credential could be implemented in an external library, instead of
+including it in gRPC core. However, we would like to use grpc_cli,
+which is part of gRPC core, with GSSAPI authentication.
```

Comment:
If we use this external library approach, is it possible to build GSSAPI-enabled grpc_cli in that project? That way we maintain only one-way dependency (external library depends on grpc core) and we don't have 2 versions of grpc core.

Reply:
Right now, it's hard to build grpc_cli outside of gRPC itself; if it were possible to build grpc_cli but link it against an existing gRPC installation, instead of having to build it all at once, then the external library approach would be more attractive.